### PR TITLE
Error: skip detection: data length less than window_size

### LINKED ISSUE #690: Error: skip detection: data length less than window_size
image

## Steps to reproduce
- set `Last <n> time` as timerange
- create pattern analytic unit
- label something
- wait for detect
- refresh

### PR DESCRIPTION
**Pull request by amper43**
Fixes #690 

Imagine that user labels a large segment, two days length for example. After that user sets time period `Last 3 hour`. Server runs detection on 3 hours range that less than window size for two days segment.

## Changes
- expand range to the past for detection if it's less than intersection (intersection = 2 window sizes)
- use `timeStep` to calculate intersection if it's available

### CODE REVIEW

File: server/src/controllers/analytics_controller.ts [hunk 1]
```diff
@@ -590,6 +595,13 @@ async function getPayloadData(
   } else {
     range = await getQueryRange(analyticUnit.id, analyticUnit.detectorType);
   }
+
+  const cache = await AnalyticUnitCache.findById(analyticUnit.id);
+  const intersection = cache.getIntersection();
+  if (range.to - range.from < intersection) {
```
Comment (rozetko) [inline server/src/controllers/analytics_controller.ts]:
```suggestion
  if(range.to - range.from < intersection) {
```

File: server/spec/analytic_controller.jest.ts [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+describe('Check detection range', function() {
+    const analyticUnitObj = {
+        _id: 'test',
+        name: "test",
+        grafanaUrl: "http://127.0.0.1:3000",
+        panelId: "ZLc0KfNZk/2",
+        type: "GENERAL",
+        metric: {
+            datasource: {
```
Comment (jonyrock) [inline server/spec/analytic_controller.jest.ts]: tabs

File: server/spec/analytic_controller.jest.ts [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    };
+
+    var query = jest.fn();
+
+    it('check range >= 2 * window size', async () => {
+        const id = await prepare();
+        await runDetect(id, 2000, 3000);
+        expect(query).toBeCalledWith(id, {from: 2000, to: 4000});
+    });
+});
```
Comment (jonyrock) [inline server/spec/analytic_controller.jest.ts]: new line

File: server/spec/analytic_controller.jest.ts [hunk 4]
```diff
@@ -0,0 +1,87 @@
+import { queryByMetric } from 'grafana-datasource-kit';
+
+jest.mock('grafana-datasource-kit', () => (
+    {
```
Comment (jonyrock) [inline server/spec/analytic_controller.jest.ts]: tabs, we use 2 spaces everywhere

File: server/spec/analytic_controller.jest.ts [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+));
+
+import { saveAnalyticUnitFromObject, runDetect } from '../src/controllers/analytics_controller';
+import * as AnalyticUnit from '../src/models/analytic_units';
+import * as AnalyticUnitCache from '../src/models/analytic_unit_cache_model';
+
+import { HASTIC_API_KEY } from '../src/config';
+
+describe('Check detection range', function() {
+    const analyticUnitObj = {
```
Comment (jonyrock) [inline server/spec/analytic_controller.jest.ts]: tabs, we use 2 spaces everywhere

File: server/spec/analytic_controller.jest.ts [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+        labeledColor: "#FF99FF",
+        deletedColor: "#00f0ff",
+        detectorType: "pattern",
+        visible: true,
+        collapsed: false,
+        createdAt: {"$$date":1564476040880},
+        updatedAt: {"$$date":1564476040880}
+    }
+
+    const windowSize = 10;
```
Comment (jonyrock) [inline server/spec/analytic_controller.jest.ts]:
```suggestion
    const WINDOW_SIZW = 10;
```

File: server/spec/analytic_controller.jest.ts [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+    const timeStep = 1000;
```
Comment (jonyrock) [inline server/spec/analytic_controller.jest.ts]:
```suggestion
    const TIME_STEP = 1000;
```

**Comment by jonyrock** [comment]
... from description 
> sets time period Last 3 hour
sets for what?

**Comment by jonyrock** [comment]
@amper43 you use word "intersection", but what does it mean in this context? Please explain the word.

**Comment by amper43** [comment]
@jonyrock 
>sets for what?
time range in grafana dashboard
image
>you use word "intersection", but what does it mean in this context? Please explain the word.
intersection is a minimal time range, that can contain segment. This word come from detection spans, there it mean intersection of spans for ability of detection on borders.
image

File: analytics/analytics/detectors/pattern_detector.py [hunk 8]
```diff
@@ -48,8 +48,9 @@ def __init__(self, pattern_type: str, analytic_unit_id: AnalyticUnitId):
 
     def train(self, dataframe: pd.DataFrame, segments: List[dict], cache: Optional[ModelCache]) -> ModelCache:
         # TODO: pass only part of dataframe that has segments
-        self.model.state = self.model.get_state(cache)
-        new_cache = self.model.fit(dataframe, segments, self.analytic_unit_id)
+        self.model.state: models.ModelState = self.model.get_state(cache)
+        new_cache: models.ModelState = self.model.fit(dataframe, segments, self.analytic_unit_id)
+        new_cache.time_step = utils.find_interval(dataframe)
```
Comment (jonyrock) [inline analytics/analytics/detectors/pattern_detector.py]: are you sure that `find_interval` always returns something useful? maybe some check here?

File: analytics/analytics/detectors/pattern_detector.py [hunk 9]
```diff
@@ -75,7 +76,7 @@ def detect(self, dataframe: pd.DataFrame, cache: Optional[ModelCache]) -> Detect
             raise ValueError(message)
 
         if len(dataframe) < window_size * 2:
-            message = f'{self.analytic_unit_id} skip detection: data length: {len(dataframe)} less than window_size: {window_size}'
+            message = f'{self.analytic_unit_id} skip detection: data length: {len(dataframe)} less than window_size: {window_size * 2}'
```
Comment (jonyrock) [inline analytics/analytics/detectors/pattern_detector.py]: `less than window_size: {window_size * 2}` looks suspicious.  maybe `less than window_size * 2: {window_size * 2}` ?

Comment (jonyrock) [reply analytics/analytics/detectors/pattern_detector.py]: btw, does user really understand the idea of "2 * window_size"?

File: server/spec/analytic_controller.jest.ts [hunk 10]
```diff
@@ -0,0 +1,89 @@
+import { queryByMetric } from 'grafana-datasource-kit';
+
+jest.mock('grafana-datasource-kit', () => (
```
Comment (jonyrock) [inline server/spec/analytic_controller.jest.ts]:
```suggestion

jest.mock('grafana-datasource-kit', () => (
```

File: server/spec/analytic_controller.jest.ts [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+  }
+));
+
+import { saveAnalyticUnitFromObject, runDetect } from '../src/controllers/analytics_controller';
+import * as AnalyticUnit from '../src/models/analytic_units';
+import * as AnalyticUnitCache from '../src/models/analytic_unit_cache_model';
+
+import { HASTIC_API_KEY } from '../src/config';
+
+describe('Check detection range', function() {
```
Comment (jonyrock) [inline server/spec/analytic_controller.jest.ts]:
```suggestion

describe('Check detection range', function() {
```

File: server/spec/analytic_controller.jest.ts [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+    visible: true,
+    collapsed: false,
+    createdAt: {"$$date":1564476040880},
+    updatedAt: {"$$date":1564476040880}
+  }
+
+  const WINDOW_SIZE = 10;
+  const TIME_STEP = 1000;
+
+  async function prepare(): Promise<string> {
```
Comment (jonyrock) [inline server/spec/analytic_controller.jest.ts]: I would suggest to choose another name. Not sure I understand the purpose of the function from the name `prepare`

File: server/src/models/analytic_unit_cache_model.ts [hunk 13]
```diff
@@ -35,7 +35,11 @@ export class AnalyticUnitCache {
   public getIntersection(): number {
     if(_.has(this.data, 'windowSize')) {
       //TODO: return one window size after resolving https://github.com/hastic/hastic-server/issues/508
-      return this.data.windowSize * 2 * MILLISECONDS_IN_INDEX;
+      if(_.has(this.data, 'timeStep')) {
```
Comment (jonyrock) [inline server/src/models/analytic_unit_cache_model.ts]: why not `this.data, 'timeStep' === undefined` ?

**Comment by jonyrock** [comment]
@amper43  it it possible if pattern be detected twice because of close intersections? 

For example, you take a small timerange and then take a small timerange a bit "righter".
Looks like "expanding" will produce very close "intersections".

**Comment by amper43** [comment]
@jonyrock 
>it it possible if pattern be detected twice because of close intersections?

duplication will be discarded

File: analytics/analytics/detectors/pattern_detector.py [hunk 14]
```diff
@@ -48,8 +48,13 @@ def __init__(self, pattern_type: str, analytic_unit_id: AnalyticUnitId):
 
     def train(self, dataframe: pd.DataFrame, segments: List[dict], cache: Optional[ModelCache]) -> ModelCache:
         # TODO: pass only part of dataframe that has segments
-        self.model.state = self.model.get_state(cache)
-        new_cache = self.model.fit(dataframe, segments, self.analytic_unit_id)
+        self.model.state: models.ModelState = self.model.get_state(cache)
+        new_cache: models.ModelState = self.model.fit(dataframe, segments, self.analytic_unit_id)
+
+        #time step isn't mandatory
```
Comment (jonyrock) [inline analytics/analytics/detectors/pattern_detector.py]:
```suggestion
        #time step is optional
```

File: analytics/analytics/detectors/pattern_detector.py [hunk 15]
```diff
@@ -75,7 +80,7 @@ def detect(self, dataframe: pd.DataFrame, cache: Optional[ModelCache]) -> Detect
             raise ValueError(message)
 
         if len(dataframe) < window_size * 2:
-            message = f'{self.analytic_unit_id} skip detection: data length: {len(dataframe)} less than window_size: {window_size}'
+            message = f'{self.analytic_unit_id} skip detection: dataset length {len(dataframe)} points less than minimal length {window_size * 2} points'
```
Comment (jonyrock) [inline analytics/analytics/detectors/pattern_detector.py]: What do you think if we make a special type of python exception (error) which sounds something like `TooSmallDataset` and improve our practice working with exceptions in the project.

File: analytics/analytics/models/model.py [hunk 16]
```diff
@@ -68,6 +68,7 @@ class ModelState():
```
Comment (jonyrock) [inline analytics/analytics/models/model.py]: please add doc comment what is the purpose of this class

File: server/spec/analytic_controller.jest.ts [hunk 17]
```diff
@@ -0,0 +1,90 @@
...
+    ...(jest.requireActual('grafana-datasource-kit')),
+    queryByMetric: jest.fn((metric, url, from, to, apiKey) => {})
+  }
+));
+
+import { saveAnalyticUnitFromObject, runDetect } from '../src/controllers/analytics_controller';
+import * as AnalyticUnit from '../src/models/analytic_units';
+import * as AnalyticUnitCache from '../src/models/analytic_unit_cache_model';
+
+import { HASTIC_API_KEY } from '../src/config';
```
Comment (jonyrock) [inline server/spec/analytic_controller.jest.ts]: I think you don't need to import a real `HASTIC_API_KEY` 
The idea is to make tests as small (as "unit") as possible. What do you think?

File: server/spec/analytic_controller.jest.ts [hunk 18]
```diff
@@ -0,0 +1,90 @@
...
+    collapsed: false,
+    createdAt: {"$$date":1564476040880},
+    updatedAt: {"$$date":1564476040880}
+  }
+
+  const WINDOW_SIZE = 10;
+  const TIME_STEP = 1000;
+
+  async function addTestUnitToDB(): Promise<string> {
+    const analyticUnitId = await saveAnalyticUnitFromObject(analyticUnitObj);
```
Comment (jonyrock) [inline server/spec/analytic_controller.jest.ts]: do you do actual saving in test?

File: server/spec/analytic_controller.jest.ts (hunk 17 above)
Comment (amper43) [reply server/spec/analytic_controller.jest.ts]: this isn't real api key. it mocked in `spec/setup_tests.ts`

File: analytics/analytics/models/model.py (hunk 16 above)
Comment (amper43) [reply analytics/analytics/models/model.py]: lets do this in other PR

File: server/spec/analytic_controller.jest.ts (hunk 18 above)
Comment (amper43) [reply server/spec/analytic_controller.jest.ts]: no, it is fake saving
